src/ws/server.rs

```diff
@@ -21,476 +21,134 @@
-
-		Ok(RegisteredNotification { queries_rx: rx })
-	}
-
-	/// Registers a method towards the server.
-	///
-	/// Clients will then be able to call this method.
-	/// The returned object allows you to handle incoming requests.
-	///
-	/// Contrary to [`register_notifications`](Server::register_notifications), there is no
-	/// `allow_losses` parameter here. If the handler is too slow to process requests, then the
-	/// server automatically returns an "internal error" to the client.
-	///
-	/// Returns an error if the method name was already registered.
-	pub fn register_method(&self, method_name: String) -> Result<RegisteredMethod, Error> {
-		if !self.registered_methods.lock().insert(method_name.clone()) {
-			return Err(Error::MethodAlreadyRegistered(method_name));
-		}
-
-		log::trace!("[frontend]: register_method={}", method_name);
-		let (tx, rx) = mpsc::channel(32);
-
-		self.to_back
-			.unbounded_send(FrontToBack::RegisterMethod { name: method_name, handler: tx })
-			.map_err(|e| Error::Internal(e.into_send_error()))?;
-
-		Ok(RegisteredMethod { to_back: self.to_back.clone(), queries_rx: rx })
-	}
-
-	/// Registers a subscription towards the server.
-	///
-	/// Clients will then be able to call this method.
-	/// The returned object allows you to send out notifications.
-	///
-	/// Returns an error if one of the method names was already registered.
-	pub fn register_subscription(
-		&self,
-		subscribe_method_name: String,
-		unsubscribe_method_name: String,
-	) -> Result<RegisteredSubscription, Error> {
-		{
-			let mut registered_methods = self.registered_methods.lock();
-
-			// NOTE: We optimise for the case where neither method is registered.
-			// This means that if the strings are equal this will be slower than just comparing the
-			// strings.
-			if !registered_methods.insert(subscribe_method_name.clone()) {
-				return Err(Error::MethodAlreadyRegistered(subscribe_method_name));
-			}
-			if !registered_methods.insert(unsubscribe_method_name.clone()) {
-				registered_methods.remove(&subscribe_method_name);
-				return Err(Error::MethodAlreadyRegistered(unsubscribe_method_name));
-			}
-		}
-
-		log::trace!(
-			"[frontend]: server register subscription: subscribe_method={}, unsubscribe_method={}",
-			subscribe_method_name,
-			unsubscribe_method_name
-		);
-		let unique_id = self.next_subscription_unique_id.fetch_add(1, atomic::Ordering::Relaxed);
-
-		self.to_back
-			.unbounded_send(FrontToBack::RegisterSubscription {
-				unique_id,
-				subscribe_method: subscribe_method_name,
-				unsubscribe_method: unsubscribe_method_name,
-			})
-			.map_err(|e| Error::Internal(e.into_send_error()))?;
-
-		Ok(RegisteredSubscription { to_back: self.to_back.clone(), unique_id })
-	}
+    pub fn register_method<F, R>(&mut self, method_name: &'static str, callback: F)
+    where
+        R: Serialize,
+        F: Fn(&str) -> Result<R, RpcError> + Send + Sync + 'static, // TODO: figure out correct lifetime here
```

Comment:
`Methods` type needs an explicit lifetime for that to work, seems like a dead-end because `fn start` calls `tokio::spawn`